src/Microsoft.IdentityModel.JsonWebTokens/Json/JsonClaimSet.cs

```diff
@@ -0,0 +1,414 @@
+                            else
+                                return JsonSerializer.Deserialize<T>(jsonElement.GetRawText());
+
+                        if (typeof(T) == typeof(double) && double.TryParse(jsonElement.ToString(), out double dresult))
+                            return (T)(object)dresult;
+
+                        if (typeof(T) == typeof(float) && float.TryParse(jsonElement.ToString(), out float fresult))
+                            return (T)(object)fresult;
+                    }
+
+                    return JsonSerializer.Deserialize<T>(jsonElement.GetRawText());
+                }
+            }
+            catch (Exception ex)
+            {
+                found = false;
+                if (throwEx)
+                    throw LogHelper.LogExceptionMessage(new ArgumentException(LogHelper.FormatInvariant(LogMessages.IDX14305, key, typeof(T), jsonElement.ValueKind, jsonElement.GetRawText()), ex));
+            }
+
+            return default;
+        }
+
+        internal bool TryGetClaim(string key, string issuer, out Claim claim)
+        {
+            if (!RootElement.TryGetProperty(key, out JsonElement jsonElement))
+            {
+                claim = null;
+                return false;
+            }
+
+            claim = CreateClaimFromJsonElement(key, issuer, jsonElement);
+            return true;
+        }
+
+        /// <summary>
+        /// The return types that are expected in a JWT token.
+        /// The 5 basic types: number, string, true / false, nil, array (of basic types).
+        /// This is not a general purpose translation layer for complex types.
+        /// For that we would need to provide a way to hook a JsonConverter to for complex types.
+        /// </summary>
+        /// <typeparam name="T"></typeparam>
+        /// <param name="key"></param>
+        /// <param name="value"></param>
+        /// <returns></returns>
+        internal bool TryGetValue<T>(string key, out T value)
+        {
+            value = GetValue<T>(key, false, out bool found);
+            return found;
+        }
+
+        internal bool HasClaim(string claimName)
+        {
+            return RootElement.TryGetProperty(claimName, out _);
+        }
+
+        private static long ParseTimeValue(string claimName, JsonElement jsonElement)
+        {
+            if (jsonElement.ValueKind == JsonValueKind.Number)
+            {
+                if (jsonElement.TryGetInt64(out long retValLong))
+                    return retValLong;
+
+                if (jsonElement.TryGetDouble(out double retValDouble))
+                    return (long)retValDouble;
+
+                if (jsonElement.TryGetInt32(out int retValInt))
+                    return retValInt;
+
+                if (jsonElement.TryGetDecimal(out decimal retValDecimal))
+                    return (long)retValDecimal;
+            }
+
+            if (jsonElement.ValueKind == JsonValueKind.String)
+            {
```

Comment:
nit: Maybe check for String type first then Number? By default a DateTime is created according to the ISO 8601-1:2019 format, which seems to be string format like "2019-07-26T16:59:57-05:00".